### PR TITLE
OpenCV & conan

### PR DESCRIPTION
**Pull request by SSE4**
/cc @conan-io/barbarians @conan-io/community-leaders

### CODE REVIEW

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 1]
```diff
@@ -0,0 +1,145 @@
+---
+layout: post
+comments: false
+title: "OpenCV conan packages: options, dependencies and challenges"
+description: "OpenCV conan packages are available in conan-center, but there are still things to be done"
+---
+
+### What is OpenCV?
+
+If you're not familar with OpenCV yet, check out our previous blog-post about [OpenCV 4.0.0](https://blog.conan.io/2018/12/19/New-OpenCV-release-4-0.html).
```
Comment (lasote) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]: familiar

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+Given the fact OpenCV is huge library with lots of features for various use-case, it's good example to demonstrate some typical package challenges (and probably few more specific as well).
```
Comment (lasote) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]: a huge

Comment (lasote) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: it's a

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+### OpenCV conan packages
+
+Recently, we have finally accepted OpenCV recipe into [conan-center](https://bintray.com/conan-community/conan/opencv%3Aconan).
+We support all major releases, so we have the following version available on bintray:
+
+* 4.x - [opencv/4.0.1@conan/stable](https://bintray.com/conan-community/conan/opencv%3Aconan/4.0.1%3Astable)
+* 3.x - [opencv/3.4.5@conan/stable](https://bintray.com/conan-community/conan/opencv%3Aconan/3.4.5%3Astable)
+* 2.x - [opencv/2.4.13.5@conan/stable](https://bintray.com/conan-community/conan/opencv%3Aconan/2.4.13.5%3Astable)
+
+Installation with conan should be pretty straighforward, e.g. you may use the following [conanfile.txt](https://docs.conan.io/en/latest/reference/conanfile_txt.html) to consume OpenCV 4.0.0:
```
Comment (lasote) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]: straightforward

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+[generators]
+cmake
+
+{% endhighlight %}
+
+As usual, pre-built packages are available for major platforms (Windows/Linux/MacOS) and compilers (Visual Studio/GCC/Clang).
+
+### dependencies
+
+OpenCV is very complex library, and has lots of various dependencies. Current conan recipe has the following:
```
Comment (lasote) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]: a very

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+<p class="centered">
+    <img  src="{{ site.url }}/assets/post_images/2019-01-21/opencv-deps.png"  align="center"
+    width="800"  alt="dependencies of OpenCV 4.0.0 package"/>
+</p>
+
+Graph was generated by the [conan info](https://docs.conan.io/en/latest/reference/commands/consumer/info.html) command:
```
Comment (lasote) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]: A graph

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+[options]
+opencv:contrib=True
+
+[generators]
+cmake
+
+{% endhighlight %}
+
+### Future : other options and dependencies
```
Comment (lasote) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]: Future:

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+As stated previously, OpenCV is very large and complex library, and it really has tons of options. And currently, our conan package doesn't support them all. You may check list of available options on their [GitHub repository](https://github.com/opencv/opencv/blob/4.0.1/CMakeLists.txt#L208). It literally takes almost 300 lines of CMake code just to declare all these options! This is something that actually hard to model in one shot. Moreover, most of options depend on other 3rd-party libraries.
```
Comment (lasote) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]: a very large

Comment (lasote) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: check the list

Comment (lasote) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: most of the options

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+Just few examples:
```
Comment (lasote) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]: a few

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+* [Google Protocol Buffers](https://developers.google.com/protocol-buffers/) might be used to read data from [Caffe](http://caffe.berkeleyvision.org/) networks
+* [OpenCL](https://www.khronos.org/opencl/) and [CUDA](https://developer.nvidia.com/cuda-zone) might be used to accelerate OpenCV algorithms on [heterogenous systems](https://en.wikipedia.org/wiki/Heterogeneous_System_Architecture)
```
Comment (lasote) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]: heterogeneous

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+* [FFMPEG](https://github.com/bincrafters/conan-ffmpeg) and [GStreamer](https://gstreamer.freedesktop.org/) might be used to read and write video files
+
+#### Google Protocol Buffers (Protobuf)
+
+OpenCV module [DNN](https://docs.opencv.org/4.0.1/d6/d0f/group__dnn.html) (Deep Neural Network) may be compiled with [Google Protobuf](https://developers.google.com/protocol-buffers/) support.
+
+We're currently actively working on adding Google Protobuf [recipe](https://github.com/bincrafters/conan-protobuf) accepted into [conan-center](https://bintray.com/conan-community/conan/opencv%3Aconan). Library is itself challenging, especially for cross-compilation use case. As soon as it's accepted, we are going to enable our OpenCV package to use Protocol Buffers by default.
```
Comment (lasote) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]: The library

Comment (lasote) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: the cross-compilation use case

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+* Windows needs SDK package provided by one of the vendors (e.g. from [Intel](https://software.intel.com/en-us/intel-opencl-support) or from [nVidia](https://developer.nvidia.com/cuda-toolkit))
+* Android also needs SDK package from vendors (e.g. [Mali](https://developer.arm.com/products/software/mali-sdks))
+
+Therefore, in order to provide OpenCL support for OpenCV package, we need to develop a way how to model such kind of dependency. Probably, we may decide to add a new feature to support so-called [virtual packages](https://www.debian.org/doc/manuals/debian-faq/ch-pkg_basics.en.html#s-virtual).
+
+#### CUDA
+
+Similar story to OpenCL, however, there is only one vendor, obviously - building of [CUDA](https://developer.nvidia.com/cuda-zone) applications requires [nVidia CUDA Toolkit](https://developer.nvidia.com/cuda-toolkit). Toolkit is pretty large, and contains CUDA compiler, in addition to libraries and headers. We either have to require user to have CUDA installed on machine during build, or provide a package for toolkit.
```
Comment (lasote) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]: the building

Comment (lasote) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: The toolkit

Comment (lasote) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: the CUDA

Comment (lasote) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: the machine

Comment (lasote) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: the build

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+#### GStreamer
+
+We're currently working on packaging [GStramer](https://gstreamer.freedesktop.org/) libraries. Similarly to FFMPEG and Google Protobuf, GStreamer itself is pretty large, and requires few other libraries to be packaged first, such as [libffi](https://sourceware.org/libffi/) and [GLib](https://developer.gnome.org/glib/stable/). Along with FFMPEG, GStreamer is one of top-requested libraries to be packaged in conan, and it's obviously on our radar.
+
+
+### Conclusion
+
+Although OpenCV packages are available in [conan-center](https://bintray.com/conan-community/conan/opencv%3Aconan), they aren't complete in term of supported options and dependencies, and we are looking into adding more in small iterations, in order to satisfy more use-cases and support more features.
+
+But we still encourage users to try our OpenCV packages, and report any issues and feature requests to our [GitHub](https://github.com/conan-community/community/issues). We will be adding missing pieces prioritizing them according to the feedback.
```
Comment (lasote) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]: I miss here some more Conan goods promotion. Some ideas:

- With Conan, you can manage N different binaries for many different flavors and platforms. With other package managers, you get "the binaries" for windows. Conan truly models the different configurations for a library at a binary level. So you can share the right configuration with your team in your company.
- No matter how complex is a library or how many different flavors or platforms do you want to support, you can write a Conan recipe to support them all or only the ones you need.

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 13]
```diff
@@ -0,0 +1,145 @@
+---
+layout: post
+comments: false
+title: "OpenCV conan packages: options, dependencies and challenges"
+description: "OpenCV conan packages are available in conan-center, but there are still things to be done"
```
Comment (lasote) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]: I don't like to transmit the idea of an incomplete package. The package is wide more complete than other package managers.

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 14]
```diff
@@ -0,0 +1,145 @@
+---
+layout: post
+comments: false
+title: "OpenCV conan packages: options, dependencies and challenges"
```
Comment (lasote) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]: OpenCV Conan package making off and future challenges?

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+#### OpenCL
+
+OpenCV may also be configured to use [OpenCL](https://www.khronos.org/opencl/), however, its support is very different across various platforms, for instance:
+
+* MacOS has built-in OpenCL support by providing *OpenCL.framework*
+* Linux needs installation of development packages (e.g. [ocl-icd-opencl-dev](https://packages.debian.org/stretch/ocl-icd-opencl-dev) on Debian systems)
+* Windows needs SDK package provided by one of the vendors (e.g. from [Intel](https://software.intel.com/en-us/intel-opencl-support) or from [nVidia](https://developer.nvidia.com/cuda-toolkit))
+* Android also needs SDK package from vendors (e.g. [Mali](https://developer.arm.com/products/software/mali-sdks))
+
+Therefore, in order to provide OpenCL support for OpenCV package, we need to develop a way how to model such kind of dependency. Probably, we may decide to add a new feature to support so-called [virtual packages](https://www.debian.org/doc/manuals/debian-faq/ch-pkg_basics.en.html#s-virtual).
```
Comment (jgsogo) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]: I understand the concept, but I'm not sure if this `virtual packages` feature is in the roadmap right now.

File: _posts/2019-01-21-OpenCV-conan-packages.markdown (hunk 1 above)
Comment (SSE4) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: done

File: _posts/2019-01-21-OpenCV-conan-packages.markdown (hunk 3 above)
Comment (SSE4) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: done

File: _posts/2019-01-21-OpenCV-conan-packages.markdown (hunk 4 above)
Comment (SSE4) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: done

File: _posts/2019-01-21-OpenCV-conan-packages.markdown (hunk 5 above)
Comment (SSE4) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: done

File: _posts/2019-01-21-OpenCV-conan-packages.markdown (hunk 2 above)
Comment (SSE4) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: done

File: _posts/2019-01-21-OpenCV-conan-packages.markdown (hunk 6 above)
Comment (SSE4) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: done

File: _posts/2019-01-21-OpenCV-conan-packages.markdown (hunk 7 above)
Comment (SSE4) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: done

File: _posts/2019-01-21-OpenCV-conan-packages.markdown (hunk 9 above)
Comment (SSE4) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: done

File: _posts/2019-01-21-OpenCV-conan-packages.markdown (hunk 8 above)
Comment (SSE4) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: done

File: _posts/2019-01-21-OpenCV-conan-packages.markdown (hunk 10 above)
Comment (SSE4) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: done

File: _posts/2019-01-21-OpenCV-conan-packages.markdown (hunk 11 above)
Comment (SSE4) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: done

File: _posts/2019-01-21-OpenCV-conan-packages.markdown (hunk 14 above)
Comment (SSE4) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: done

File: _posts/2019-01-21-OpenCV-conan-packages.markdown (hunk 13 above)
Comment (SSE4) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: do you have suggestion how to re-phrase it?

File: _posts/2019-01-21-OpenCV-conan-packages.markdown (hunk 15 above)
Comment (SSE4) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: I am not saying it's on roadmap, I am just saying we may decide to add it, or may not. but if you feel it shouldn't be here, I can remove it.

**Comment by SSE4** [comment]
I've added some sections which describe details on how to write build method for OpenCV and how to manage requirements.
I am still thinking on lessons part, if you have some ideas, please tell me :)

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 16]
```diff
@@ -0,0 +1,244 @@
...
+title: "OpenCV Conan package making off and future challenges"
+description: "OpenCV conan packages are available in conan-center, but there are still things to be done"
+---
+
+### What is OpenCV?
+
+If you're not familiar with OpenCV yet, check out our previous blog-post about [OpenCV 4.0.0](https://blog.conan.io/2018/12/19/New-OpenCV-release-4-0.html).
+Given the fact OpenCV is a huge library with lots of features for various use-case, it's a good example to demonstrate some typical package challenges (and probably few more specific as well).
+
+### OpenCV conan packages
```
Comment (danimtb) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]:
```suggestion
### OpenCV's Conan packages
```

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 17]
```diff
@@ -0,0 +1,244 @@
...
+
+Recently, we have finally accepted OpenCV recipe into [conan-center](https://bintray.com/conan-community/conan/opencv%3Aconan).
+We support all major releases, so we have the following version available on bintray:
```
Comment (danimtb) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]:
```suggestion
We support all major releases, so we have the following version available on Bintray:
```

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 18]
```diff
@@ -0,0 +1,244 @@
...
+
+* 4.x - [opencv/4.0.1@conan/stable](https://bintray.com/conan-community/conan/opencv%3Aconan/4.0.1%3Astable)
+* 3.x - [opencv/3.4.5@conan/stable](https://bintray.com/conan-community/conan/opencv%3Aconan/3.4.5%3Astable)
+* 2.x - [opencv/2.4.13.5@conan/stable](https://bintray.com/conan-community/conan/opencv%3Aconan/2.4.13.5%3Astable)
+
+Installation with conan should be pretty straightforward, e.g. you may use the following [conanfile.txt](https://docs.conan.io/en/latest/reference/conanfile_txt.html) to consume OpenCV 4.0.0:
```
Comment (danimtb) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]:
```suggestion
Installation with Conan should be pretty straightforward, e.g. you may use the following [conanfile.txt](https://docs.conan.io/en/latest/reference/conanfile_txt.html) to consume OpenCV 4.0.0:
```

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 19]
```diff
@@ -0,0 +1,244 @@
...
+opencv/4.0.1@conan/stable
+
+[generators]
+cmake
+
+{% endhighlight %}
+
+As usual, pre-built packages are available for major platforms (Windows/Linux/MacOS) and compilers (Visual Studio/GCC/Clang).
+
+### building OpenCV
```
Comment (danimtb) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]:
```suggestion
### Building OpenCV
```

File: _posts/2019-01-21-OpenCV-conan-packages.markdown (hunk 13 above)
Comment (danimtb) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: something like "Hands-on packaging OpenCV (the proper way)". May sound a bit presumptuous and we could omit the last part but I think this is a blog post to show how to do it when packaging big and reasonably tricky libraries

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 20]
```diff
@@ -0,0 +1,244 @@
...
+OpenCV uses [CMake](https://cmake.org/), therefore our [recipe](https://github.com/conan-community/conan-opencv/blob/release/4.0.1/conanfile.py) uses [CMake build helper](https://docs.conan.io/en/latest/reference/build_helpers/cmake.html#cmake-reference). The process to build CMake-based project is typical for many recipes, and OpenCV is not an exception here.
+
+First step is to configure CMake:
+
+{% highlight python %}
+
+    def _configure_cmake(self):
+        cmake = CMake(self)
+        cmake.definitions['BUILD_EXAMPLES'] = False
+        # configure various OpenCV options via cmake.definitions
```
Comment (danimtb) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]: Do not omit the interesting things. I would like to see explanations for newbies but also complex things even explainng each opencv option that affects the build

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 21]
```diff
@@ -0,0 +1,244 @@
...
+{% highlight python %}
+
+    def build(self):
+        # intentionally skipped code to patch OpenEXR here
+        cmake = self._configure_cmake()
+        cmake.build()
+
+{% endhighlight %}
+
+*cmake.build()* executes build tool depending on CMake [generator](https://cmake.org/cmake/help/v3.13/manual/cmake-generators.7.html), it might be [MSBuild](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild?view=vs-2017), [GNU Make](https://www.gnu.org/software/make/), [Ninja](https://ninja-build.org/), etc.
```
Comment (danimtb) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]: Here you can also say that this is one of the reasons (not the only one) that makes CMake is so convenient

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 22]
```diff
@@ -0,0 +1,244 @@
...
+{% highlight python %}
+
+    def package(self):
+        cmake = self._configure_cmake()
+        cmake.install()
+        cmake.patch_config_paths()
+
+{% endhighlight %}
+
+It doesn't have typical code to copy platform-specific files, like .dll, .so, .dylib, etc. Instead, it uses CMake [install](https://cmake.org/cmake/help/v3.13/command/install.html) feature. CMake may generate special target called *INSTALL*, which copies project's header, libraries, CMake [configuration](https://cmake.org/cmake/help/v3.13/manual/cmake-packages.7.html) files, [pkg-config](https://www.freedesktop.org/wiki/Software/pkg-config/) files, other data files, like [Haar Cascades](https://docs.opencv.org/4.0.1/d7/d8b/tutorial_py_face_detection.html) in case of OpenCV. So, if project itself knows which files to distribute and how to properly layout them, then it doesn't make much sense to replicate this logic in conanfile, right?
```
Comment (danimtb) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]: Would be interesting to explain here that ``cmake.install()`` helpers already makes the proper adjustments to install everything in the package folder

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 23]
```diff
@@ -0,0 +1,244 @@
...
+
```
Comment (danimtb) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]: This is a great point!! It is exactly what I am expecting to read en a post like this one 👍

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 24]
```diff
@@ -0,0 +1,244 @@
...
+But what is *cmake.patch_config_paths()* and why do we need it? Well, CMake-generated config files may contain absolute paths, which something we would like to avoid, because such paths are specific to the machine where recipe was built, and consumers usually won't have dependencies installed in the same paths. For instance, on Windows conan directory usually located within [USERPROFILE](https://msdn.microsoft.com/en-us/library/windows/desktop/bb776892(v=vs.85).aspx) directory, which contains user name (e.g. *AppVeyor*). Given that fact, usage of generated CMake config files may result in inability to build project, so there is a workaround for this problem in conan.
+
+### dependencies
+
+OpenCV is a very complex library, and has lots of various dependencies. Current conan recipe has the following:
```
Comment (danimtb) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]:
```suggestion
OpenCV is a very complex library, and has lots of various dependencies. Current Conan recipe has the following:
```

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 25]
```diff
@@ -0,0 +1,244 @@
...
+
+{% highlight bash %}
+
+$ conan info opencv/4.0.0@conan/stable --graph opencv
+
+{% endhighlight %}
+
+As you can see, currently it mostly depends on image libraries, such as [libjpeg](http://libjpeg.sourceforge.net/), [libtiff](http://www.libtiff.org/), [libpng](http://www.libpng.org/pub/png/libpng.html), [libwepb](https://developers.google.com/speed/webp/), [jasper](http://www.ece.uvic.ca/~frodo/jasper/) and [OpenEXR](http://www.openexr.com/).
+
+All these libraries are available as conan packages in [conan-center](https://bintray.com/conan-community/conan/opencv%3Aconan) as well.
```
Comment (danimtb) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]:
```suggestion
All these libraries are available as Conan packages in [conan-center](https://bintray.com/conan-community/conan/opencv%3Aconan) as well.
```

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 26]
```diff
@@ -0,0 +1,244 @@
...
+
+These libraries are mainly needed by OpenCV [imgcodecs](https://docs.opencv.org/4.0.1/d4/da8/group__imgcodecs.html), to support reading and writing of various image formats.
+
+All mentioned libraries might be enabled or disabled using conan options (They are currently enabled by default). For instance, to disable [OpenEXR](http://www.openexr.com/) support, use the following:
```
Comment (danimtb) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]:
```suggestion
All mentioned libraries might be enabled or disabled using options (They are currently enabled by default). For instance, to disable [OpenEXR](http://www.openexr.com/) support, use the following:
```

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 27]
```diff
@@ -0,0 +1,244 @@
...
+
+### OpenCV contrib
+
+In addition to the built-in features, OpenCV has a collection of extra modules, called [OpenCV contrib](https://github.com/opencv/opencv_contrib). Currently, it has about 100 additional modules! Just o name a few:
+
+* [Face Analysis](https://docs.opencv.org/4.0.1/db/d7c/group__face.html)
+* [Optical Flow](https://docs.opencv.org/4.0.1/d2/d84/group__optflow.html)
+* [Image Registration](https://docs.opencv.org/4.0.1/db/d61/group__reg.html)
+
+By default, our package doesn't have OpenCV contrib modules enabled. But you may easily have them available by passing *opencv:contrib* option:
```
Comment (danimtb) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]: Explain how this option affect the package: Is this additional source code? maybe enable some option in cmake? does it add new requirements?

File: _posts/2019-01-21-OpenCV-conan-packages.markdown (hunk 12 above)
Comment (danimtb) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: Completely agree with @lasote

File: _posts/2019-01-21-OpenCV-conan-packages.markdown (hunk 21 above)
Comment (SSE4) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: good point, it's worth adding

File: _posts/2019-01-21-OpenCV-conan-packages.markdown (hunk 13 above)
Comment (SSE4) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: okay, I've changed to the suggested description

File: _posts/2019-01-21-OpenCV-conan-packages.markdown (hunk 22 above)
Comment (SSE4) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: added explanation

File: _posts/2019-01-21-OpenCV-conan-packages.markdown (hunk 21 above)
Comment (SSE4) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: added

File: _posts/2019-01-21-OpenCV-conan-packages.markdown (hunk 20 above)
Comment (SSE4) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: I've added another paragraphs explaining these interesting details

File: _posts/2019-01-21-OpenCV-conan-packages.markdown (hunk 27 above)
Comment (SSE4) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: added explanation

**Comment by SSE4** [comment]
I am adding your comments with check boxes here to simplify tracking (it's easier for me):

- [ ] With Conan, you can manage N different binaries for many different flavors and platforms. With other package managers, you get "the binaries" for windows. Conan truly models the different configurations for a library at a binary level. So you can share the right configuration with your team in your company.
- [ ] No matter how complex is a library or how many different flavors or platforms do you want to support, you can write a Conan recipe to support them all or only the ones you need.

- [ ] How to define requirements to other conan packages.
- [ ]How the integration and usage of those packages is done: is it necessary to change the build scripts? what generators are being used?
- [ ] What is special about building OpenCV? How the build() method looks like.
- [ ] Lessons learnt in the recipe.

- [ ] First, I think the focus of this post should be showing people how to package a complex library.
- [ ]Show points on why the usage and distribution of opencv it is not good: too many dependencies, 3rd party folders, many cmake options
- [ ]Explain useful tips and tricks for advanced packagers when creating a recipe.
- [ ] Explain the package_info() for consumers and what are the things that you should test and how you know you should provide consumers additional libraries to link with: pthread, m, Vfw32
- [ ]It is important in image libraries like this to explain system_requirements and how did you know you need them.
- [ ] Say that 3rd party folder is not needed and what options should be disabled
- [ ] Promote the fact that almost all 3rd party libraries are already in Conan Center (most of them thanks to Bincrafters too!)
- [ ] Conclude with the results: different configurations supported, times of build per package, ease of installation with Conan...
- [ ] Say that it has some pending points but do not put A LOT of emphasis on it. Say that we are willing to keep it improving and any issue/PR with help is more than welcome.

**Comment by SSE4** [comment]
I think I have added all missing things, feel free to review again :)

**Comment by lasote** [comment]
@SSE4 for the classic typos, please use Grammarly (free for most common mistakes). You have like 60 simple corrections to do.

**Comment by SSE4** [comment]
@lasote okay, I think I have fixed them

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 28]
```diff
@@ -0,0 +1,461 @@
...
+
+* Use dynamic [requirements](https://docs.conan.io/en/latest/reference/conanfile/methods.html#requirements) for optional dependencies
+* Use [build helpers](https://docs.conan.io/en/latest/reference/build_helpers.html), if possible, they automate many things and allow to keep recipe code short and clean
+* [patch_config_paths](https://docs.conan.io/en/latest/reference/build_helpers/cmake.html?highlight=patch_config_paths) might be required for CMake libraries
+* Use exelinkflags/sharedlinkflags to specify [Apple frameworks](https://docs.conan.io/en/latest/howtos/link_apple_framework.html)
+* Avoid System Requirements, if possible, package libraries with Conan instead
+
+### Conclusion
+
+Although OpenCV packages are available in [conan-center](https://bintray.com/conan-community/conan/opencv%3Aconan), they aren't complete in term of supported options and dependencies, and we are looking into adding more in small iterations, in order to satisfy more use-cases and support more features.
```
Comment (danimtb) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]: This observation is ok, but do not make it the main point of the conclusion. Conclude with something positive like: "Conan is flexible enough to achieve complex packaging" or "Conan provides tools and build helper wrappers that make packaging a lot easier" or additionally "having all the steps to build opencv in a conanfile.py is very convenient"

Use the positive points of the blogpost to enforce the idea you would like to transmit to the reader.

Comment (SSE4) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: I've added some positive points you have mentioned to the conclusion.
I've also added dots.

**Comment by SSE4** [comment]
@danimtb  Yep, you guys listed lots of things which you would like to see covered in blog-post, and that's actually lots of information, but I believe it's still good size for blog-post.

File: _posts/2019-01-21-OpenCV-conan-packages.markdown [hunk 29]
```diff
@@ -0,0 +1,465 @@
...
+                                  r'SET(OPENEXR_ROOT "C:/Deploy" CACHE STRING "Path to the OpenEXR \"Deploy\" folder")',
+                                  '')
+            tools.replace_in_file(find_openexr, r'set(OPENEXR_ROOT "")', '')
+            tools.replace_in_file(find_openexr, 'SET(OPENEXR_LIBSEARCH_SUFFIXES x64/Release x64 x64/Debug)', '')
+            tools.replace_in_file(find_openexr, 'SET(OPENEXR_LIBSEARCH_SUFFIXES Win32/Release Win32 Win32/Debug)',
+                                  '')
+
+{% endhighlight %}
+
+We use [tools.replce_in_file](https://docs.conan.io/en/latest/reference/tools.html?highlight=replace_in_file#tools-replace-in-file) here to remove several lines of CMake code. In more complex cases, [tools.patch](https://docs.conan.io/en/latest/reference/tools.html?highlight=patch#tools-patch) helper might be used instead.
```
Comment (memsharded) [inline _posts/2019-01-21-OpenCV-conan-packages.markdown]: typo ``[tools.replce_in_file]``

Comment (SSE4) [reply _posts/2019-01-21-OpenCV-conan-packages.markdown]: fixed

**Pull request merged**